euphonic/design/03_design_document_v0.3.md

```diff
@@ -0,0 +1,385 @@
+|Energy units|string|
+
+|Method|Input required|Output|
+|------|--------------|------|
+|Calculate crystal structure factor (see [algorithms - 2.1](04_algorithms.md))|Neutron scattering lengths, temperature, Debye-Waller exponent (optional)|Crystal structure factor (discrete energies)|
+|Calculate Debye-Waller exponent (see [algorithms - 3.2](04_algorithms.md))|Temperature|Debye-Waller exponent|
+|Calculate density of states|Energy bins, energy broadening|1D Spectrum object|
+|Serialise|Filename|File|
+
+### Crystal Structure Factor (discrete energies)
+
+This contains the structure factor calculated for a crystal at each q-point and
+for each mode. It could be passed to Brille or Horace for energy binning and
+plotting, or binned and plotted in Euphonic.
+
+|Attribute|Type|
+|---------|----|
+|Q-points|(n_qpts, 3) array of floats|
+|Structure factor|(n_qpts, 3*n_ions) array of floats|
+|Frequencies|(n_qpts, 3*n_ions) array of floats|
+|Energy units|string|
+
+|Method|Input required|Output|
+|------|--------------|------|
+|Calculate binned crystal structure factor (see [algorithms - 2.2](04_algorithms.md))|Energy bins, temperature (optional)|2D Spectrum object|
+|Serialise|Filename|File|
+
+### Debye-Waller exponent
+
+The Debye-Waller exponent is calculated as a sum over the 1st BZ, so requires
+frequencies and eigenvectors that have been calculated on a grid. It is also
+direction dependent, and resolved per ion, so will be a 3x3 array for each ion
+in the structure.
+
+It is used as an input for the structure factor calculations, this could be
+optional but is needed if the user wants to include temperature effects.
+
+|Attribute|Type|
+|---------|----|
+|Structure|Structure object|
+|Temperature|float|
+|Debye-waller|(n_ions, 3, 3) array of floats|
+|Temperature units|string|
+
+|Method|Input required|Output|
+|------|--------------|------|
+|Serialise|Filename|File
+
+### 2D Spectrum
+
+This is an object for any 2D binned spectra, and can be plotted or exported to
+another program. It is intended to be general and can be used for binned crystal
+Q-E plots, powder |Q|-E plots etc. The x-tick locations/labels have been
+included to allow helpful labelling of the x-axis, which is most often the
+Q-axis (e.g. high symmetry points), and as the structure/q-point information is
+not included in the 2D spectrum object these cannot be calculated on the fly.
+For parity it could be sensible to include y tick labels/locations, but until
+Q-Q plots are a Euphonic requirement, these are not included in the current
+object definition.
+
+|Attribute|Type|
+|---------|----|
+|x bin edges|(n_xbins + 1) array of floats|
+|x bin units|string|
+|y bin edges|(n_ybins + 1) array of floats|
+|y bin units|string|
+|x tick labels|list of strings|
+|x tick locations|array of ints|
+|plot data|(n_xbins, n_ybins) array of floats|
+
+|Method|Input required|Output|
+|------|--------------|------|
+|Plot|x,y broadening|Matplotlib Figure|
```

Comment:
I do think there should be a `.plot()` method on the data objects; it's too cumbersome to go and find the appropriate plotting function and call it on an object, especially given that these will be relatively simple plots. The method could dispatch to (otherwise-useless) code living in another module, but that's a housekeeping detail really; the point is that either way you call `plot()`, _then_ it attempts to import matplotlib and raises an ImportError if you don't have it. You can safely instantiate the object, examine it and call `serialise()` without matplotlib.

Reply:
`1dplotter.defaultplot( spectrum ) === spectrum.plot()`

Reply:
What scenario do you have in mind where that added method creates a maintenance problem? It has no impact on serialisation or at interfaces between languages, because only the stored data needs to be exchanged.

`import 1dplotter; 1dplotter.defaultplot(spectrum)` is an objectively worse user experience from an interactive shell or notebook than `spectrum.plot()`. If there's a maintenance benefit then maybe the trade-off is worth it, but I'd like to understand what it is so I can avoid the same mistake in future.

Reply:
It's not a "mistake" and I imagine counter arguments could be made for its inclusion.

I am thinking primarily of extra maintenance effort and additional cognitive load for both maintainers and users arising from the complexity of having plot functionality on the spectrum class AND the separate plotter object.

1) Plotting is be done in two places in the codebase via to distinct APIs. 
```
spectrum.plot();
```
and 
```
import plotter;
plotter.plot(spectrum);
```
This raises questions for a user which one should I be using? Is there a difference? 
For developers, if the 'default' plot is available in both there is a maintenance overhead of keeping them inline. If I want to change some style on all Euphonic plots I have to do it in the spectrum plot as well as the plotting classes.

2) The spectrum class now does two distinct things:  calculates/stores data about spectra (backend functionality) AND plots spectra, popping open a UI element to do it.
In contract the plotter class does one thing: provide (multiple) standard plots of spectrum data

3) A user  is likely have their own preferred plots and plotting libraries which they will want use for custom plots for publication or to look at data not covered by the standard set. These would need to work with an appropriately dimensioned spectrum object. By doing all the Euphonic plots via the plotter class(es) we have ensure that all data is available through the API in a usable form.

4) A lazy-import of matplotlib removes the issue of running on a machine without that library installed - if that plot routine is never called, but calling `plot` on spectrum would then fail; if you're always plotting through the plotter class that dependency becomes more explicit and your script would fail early.

The issue of having plotting routines "outside" the main analytic code is something that Greg has raised with Horace which bundles a load of plots he doesn't want in.

Reply:
Thanks for explaining, that makes sense. Redundancy is bad. My take:
- If the plotter for a 1D/2D spectrum is fairly simple then it should live on the class and not be redundantly callable from somewhere else.
- If we have a whole assortment of plotting tools for this object then it is reasonable for them to live elsewhere.

I was under the impression that the former is more accurate, as plotting is more of a convenience/diagnostic tool and not a major role of Euphonic.

> If I want to change some style on all Euphonic plots I have to do it in the spectrum plot as well as the plotting classes.

I agree with the principle here, but for this specific example it seems good to mention:
If we are doing any plot styling, we should use Matplotlib's cascading style system. Not only does it make this problem go away, it also cuts down on the need to repeat ourselves between "unrelated" plotters that can share a base stylesheet.

> 2. The spectrum class now does two distinct things: calculates/stores data about spectra (backend functionality) AND plots spectra, popping open a UI element to do it.

The spectrum classes don't do any calculating, only storage. And an object-oriented plotter naturally stores plot data, so why not have them be the same object?

Point 3 is an interesting one; basically, having an extra step where we use our own API is a good thing because it keeps us honest? We should be able to rely on testing to make sure the API behaves as advertised.

Point 4: It does suck to have a script die to surprise dependencies, so this seems a good design principle. This particular case is an *especially* unsurprising dependency; I don't think users can get mad at us because their script calling a method named `plot()` blows up without matplotlib. But there remains a theoretical risk that we will mess up by implicitly calling plot() as part of something else. If we are saying matplotlib is optional then we should make sure to have some CI that runs on a matplotlib-less container.

Reply:
Thanks both for your feedback and discussion.

As the plotting in Euphonic should be kept simple, I was leaning towards just having a plot method on the object, however there is one particular use case which will require a separate plotter module: plotting multiple 1D spectra on the same axes. This can't be done with a plot method, and we've agreed being able to plot in multiple places is a bad idea. So we'll go with the plotter module and I'll update the doc to make that clearer.